docs/usage/shoot_pod_autoscaling_best_practices.md

```diff
@@ -0,0 +1,327 @@
+- **Compatibility with HPA:** Care must be taken when using VPA in conjunction with HPA, as they can potentially interfere with each other's scaling decisions.
+
+# Combining HPA and VPA
+
+HPA and VPA serve different purposes and operate on different axes of scaling. HPA increases or decreases the number of pod replicas based on metrics like CPU or memory usage, effectively scaling the application out or in. VPA, on the other hand, adjusts the CPU and memory allocations of individual pods, scaling the application up or down.
+
+When used together, these autoscalers can provide both horizontal and vertical scaling. However, they can also conflict with each other if used on the same metric (e.g. both on CPU or both on memory). In particular, if VPA adjusts the requests, the utilization will approach 100% (for various reasons not exactly right, but for this consideration, close enough), which may trigger HPA to scale out, if it's configured to scale on utilization (often seen in simple examples), which will spread the load across more pods, which may trigger VPA again to adjust the requests. This is a feedback-loop.
+
+To effectively combine HPA and VPA, consider the following strategies:
+- **Use HPA and VPA on different metrics:** One way to avoid conflicts is to use HPA and VPA based on different metrics. For instance, you could configure HPA to scale based on requests per seconds (or another representative custom metric) and VPA to adjust CPU and memory requests. This way, each autoscaler operates independently based on its specific metric(s).
+- **Use HPA on usage, not utilization, when combining it with VPA:** HPA is often configured on `utilization` as replicas driver in simple examples, e.g. used CPU divided by requested CPU (=CPU utilization). This is a problem if VPA modifies the requested CPU as this results in a feedback loop. However, HPA also supports `usage` as replicas driver, which is absolute (requested CPU doesn't matter here). If HPA is configured with `usage` as replicas driver, you can combine HPA and VPA on the very same metric(s), if you know what you are doing.
+
+Let's zoom in on the last option with a bit more details as it is the solution that gives you the most flexibility:
+
+The primary challenge in combining HPA and VPA on the same metrics arises from the way HPA calculates the desired number of replicas. HPA uses the following formula:
+
+`desiredReplicas = ceil[currentReplicas * (currentMetricValue / desiredMetricValue)]`
+
+When VPA adjusts the resource requests for a pod, it will inadvertently affect the utilization metric used by HPA. Since utilization is a ratio of current usage to the requested resources, any change in the requests by VPA will alter the utilization as computed by HPA, potentially triggering an HPA scaling event. This can lead to a feedback loop where VPA and HPA scaling actions interfere with each other.
```

Comment:
Hmm... why does it sound similar? Line 211 explains what HPA and VPA do while line 223 explains the danger/feedback loop that is the basis of the unconditional (and therefore wrong) recommendation never to combine HPA and VPA on the same metrics (which is exactly what we now do).

So, the quality of this sentence is (should be) very different. Can you please point me to the exact words that you believe are superfluous as I have read the text too often to notice anymore, what sounds off.

Reply:
There's a few sections/parts which also seem repetitive to me. If you look up the places where "feedback" occurs in the doc, they seems to state very similar things: utilization bad, feedback loop can happen, if you want to use both on the same metrics, use average usage instead

Reply:
> In particular, if VPA adjusts the requests, the utilization will approach 100% (for various reasons not exactly right, but for this consideration, close enough), which may trigger HPA to scale out, if it's configured to scale on utilization (often seen in simple examples), which will spread the load across more pods, which may trigger VPA again to adjust the requests. This is a feedback loop.

vs 

> When VPA adjusts the resource requests for a pod, it will inadvertently affect the utilization metric used by HPA. Since utilization is a ratio of current usage to the requested resources, any change in the requests by VPA will alter the utilization as computed by HPA, potentially triggering an HPA scaling event. This can lead to a feedback loop where VPA and HPA scaling actions interfere with each other.


Both are talking about the conflict when VPA and HPA are used together. In both cases the example is when VPA adjusts pod requests which leads to change in the utilization which may trigger HPA to scale out which may trigger VPA to adjust the requests again. At the end, both states that this is the so called feedback loop. This is why I see both lines (now 221) bear similar meaning and looks like duplication to me.

Reply:
> There's a few sections/parts which also seem repetitive to me. If you look up the places where "feedback" occurs in the doc, they seems to state very similar things: utilization bad, feedback loop can happen, if you want to use both on the same metrics, use average usage instead

Yes, that's true. Multiple statements repeat when the context asks/demands for it. I don't know how else to do it or not do it, because I do not know how people will later go back to this guide (they will probably never read the document again and only reread a certain section). After all, that's not a book, but a guide, so it will be repetitive.

If you make it more concrete, where it really annoys the reader @voelzmo, I can take a look. Otherwise, repetition depending on the context is actually intended.

Reply:
> Both are talking about the conflict when VPA and HPA are used together. In both cases the example is when VPA adjusts pod requests which leads to change in the utilization which may trigger HPA to scale out which may trigger VPA to adjust the requests again. At the end, both states that this is the so called feedback loop. This is why I see both lines (now 221) bear similar meaning and looks like duplication to me.

Well, the sentences explain what the title of the section says they would do. Let me reread this section again and find a more pleasing wording, if that's too much. But please bear in mind, that you (@voelzmo and @vpnachev and all the other reviewers) know the problem/solution already, but an uninitiated end user that has never thought about that will need more time and help to get the problem/solution.

Reply:
OK, I tried to shorten/deduplicate this section and its referrers @voelzmo @vpnachev.